packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -251,28 +251,36 @@ const useOverflow = (
         // if the new average of the guess range will never change the visibleCount anymore (infinite loop)
         (maxGuess - minGuess === 1 && visibleCount === minGuess)
       ) {
-        dispatch({ type: 'stabilize' });
+        dispatch({ type: 'stabilize', state: guessState });
         return;
       }
 
       // Before the main logic, the max guess MUST be ≥ the correct visibleCount for the algorithm to work.
       // If not, should shift the guess range forward to induce the first overflow.
       if (maxGuess === visibleCount && !isOverflowing) {
-        dispatch({ type: 'shiftGuessRangeForward' });
+        dispatch({ type: 'shiftGuessRangeForward', state: guessState });
         return;
       }
 
       if (isOverflowing) {
         // overflowing = we guessed too high. So, decrease max guess
-        dispatch({ type: 'decreaseMaxGuess' });
+        dispatch({ type: 'decreaseMaxGuess', state: guessState });
       } else {
         // not overflowing = maybe we guessed too low? So, increase min guess
-        dispatch({ type: 'increaseMinGuess' });
+        dispatch({ type: 'increaseMinGuess', state: guessState });
       }
     } finally {
       isGuessing.current = false;
     }
-  }, [isStabilized, itemsCount, maxGuess, minGuess, orientation, visibleCount]);
+  }, [
+    guessState,
```

Comment:
I'll repeat what I said internally, which is that we should move `guessVisibleCount` inside the Effect, to make the execution flow more direct.

Additionally, we can make the dependency array smaller, by destructuring things from `guessState` inside the Effect.

![](https://github.com/user-attachments/assets/bbef2d48-ef03-4777-a428-a7974540634e)